# chool Districts and Funding Analysis

In [ ]:
from google.colab import files

In [ ]:
files.upload() #upload district_statistics file

In [ ]:
files.upload() #upload state_sizes file


Included are two functions for your convenience that take as input each of the two files and converts them into data structures that you can use for your code.

In [ ]:
# This function takes one object as input:
# populations_file: A filename where the file contains a spreadsheet with
#                     exactly two columns: name in column 0, population in column 1
# 
# This function returns a dictionary as output:
#   states: Each key in this dictionary is a state name.
#           The value is the population for that state, as found in the data file. 
# 
def build_state_populations_dictionary(populations_file):
    states = {}
    with open(populations_file) as states_file:
        states_input = states_file.readlines()
        header_row = []
        i = 0
        for row in states_input:
            if i == 0:
                header_row = row.split(",")
            else:
                state_row = row.split(",")
                state_name = state_row[0].strip()
                state_population = int(state_row[1].strip())
                states[state_name] = state_population
            i += 1
    return states
    
# This function takes one object as input:
#   districts_filename: A filename where the file contains a spreadsheet with
#                       any number of columns.
#
# This header column is then followed by any number of rows. The function
# attempts to convert the values to integers, or defaults to strings if it cannot.
#
# This function returns a list as output:
#   districts_list: Each entry in this list is a dictionary representing
#                   a single school district from our input file. Each dictionary
#                   contains keys corresponding to the columns in the input file. 
#                   That means you can retrieve, for instance, the 
#                   total revenue of the first school district in the list:
# 
#                           districts_list[0]["revenue"]
# 
def build_districts_list(districts_filename):
    districts_list = []
    with open(districts_filename) as districts_file:
        districts_input = districts_file.readlines()
        row_number = 0
        header_row = []
        for row in districts_input:
            if row_number == 0:
                header_row = row.strip().split(",")
            else:
                district_row = row.split(",")
                district_facts = {}
                for column_number in range(len(district_row)):
                    column_name = header_row[column_number]
                    cell_value = district_row[column_number]
                    try:
                        cell_value = int(cell_value)
                    except ValueError:
                        pass
                    district_facts[column_name] = cell_value
                districts_list.append(district_facts)
            row_number += 1
    return districts_list

In [ ]:
state_populations = build_state_populations_dictionary("state_sizes.csv")
districts_list = build_districts_list("district_statistics.csv")

# Part 1

In no less than 100 words, why might someone want to look at this data? Why would it matter?

The data we use in the analysis consists of two parts: one is the population of each state in United States, and the other is about funding and expense details of public school districts in every state with the units of 1,000 dollars. The fundiung data includes expenses in each school district, total revenues and the breakdown of funding from federal, state and local governments.

Through these two datasets, we can directly get information, and we can also go further through some mathematical formulas. We can directly know about each state, such as population, funding from all levels of governments to each public school districts, expenses each district costed, etc. Alos, we can know more using formulas and ratios, such as total funding in scale of states (adding up the data of all school districts in a state), budget deficitd of each state( gap between expenses and revenues), mean size of a school district in a state, etc. With these calculated date, we can easily make comparisons among each public school district and state and it helps readers get results and make the following decisions. 

Potential readers may be governemnt staff from education, finance, statistics and human development departments, policy makers, scholars like statisticians and sociologists, social media and so on. They can use these datasets to assess the operation of finance budgets, improve policies, do social research and make news reports. This data can give them inspriations from various aspects and help them select the most efficient methods and convincing data in their work.

Yhrough this data, data scientists can answer questions as folloes :a) financial budget questions. For example, it can answer which state received the largest amount of funding from governments and which state spent the most expenses, and which state or school district had the largest budget deficit; b) population questions, such as which state had the largest population and the average size of a school district. Also, further questions like per capita people gave, average expenses of each person in the state can also be answered; c) comparstion questions. The data can help answer differences of funding situations among states and school districts. It can make contributions to assess budget usage efficiency among atate.



# Part 2


## Q1
In our dataset, what percentage of school districts nationwide received more than $10,000,000 in local funding?


In [ ]:
def q1_ten_million(districts_list):
  count = 0  #count the number of local funding more than 10000
  sum_districts = 0   #count the total number of districts
  for item in districts_list:
    if item['local_revenue'] > 10000:
      count += 1 
      sum_districts += 1
    else:
      sum_districts += 1
  return print('The percentage of school districts nationwide received more than $10,000,000 in local funding is {:.2%}'.format(count / sum_districts))
  

In [ ]:
q1_ten_million(districts_list)

The percentage of school districts nationwide received more than $10,000,000 in local funding is 30.30%


**Conclusion:**In our dataset, about ***30.30 percentage*** of school districts nationwide received more than $10,000,000 in local funding.



## Q2

In our dataset, among school districts with a total budget (expenses) of at least $1 billion, which has the shortest name? What state is that district in?


In [ ]:
def q2_shortest_billion_name(districts_list):
  q2_list = [[item['name'],len(item['name']),item['state']] for item in districts_list if item['expenses'] > 1000000] # pick out the school districts with a total budget of at least $1 billion
  shortest_length = min(q2_list, key = lambda x:x[1]) # find the shortest length among names
  return print(shortest_length[0],'has the shortest name and the district is in',shortest_length[2])   

In [ ]:
q2_shortest_billion_name(districts_list)

DADE has the shortest name and the district is in Florida


**Conclusion:** ***DADE*** has the shortest name and the district is in ***Florida***.

##Q3

Hawaii has only one school district, which covers the entire state. Across the rest of the United States, how many school districts had greater expenses than Hawaii did statewide?


In [ ]:
def q3_larger_than_hawaii(districts_list):
  Hawaii_expense = [item['expenses'] for item in districts_list if item['state'] == 'Hawaii'] #find the expenses of Hawaii
  for i in Hawaii_expense:
    Hawaii_expense = int(i)  #change the list to integer
  count = 0
  for item in districts_list: 
    if item['expenses'] > Hawaii_expense:
      count += 1 # count the number of school districts
    else:
      pass
  return print(count,'school districts had greater expenses than Hawaii did statewide.')

In [ ]:
q3_larger_than_hawaii(districts_list)

8 school districts had greater expenses than Hawaii did statewide.


**Conclusion:** Across the rest of the United States, ***eight school districts*** had greater expenses than Hawaii did statewide. 

## Q4

The budget deficit of a school district is equal to its expenses minus its revenue. Which school district in Pennsylvania had the largest budget deficit? And how large was it?


In [ ]:
def q4_biggest_pennsylvania_district(districts_list):
  q4_list = [[item['name'],(item['expenses'] - item['revenue'])] for item in districts_list if item['state'] == 'Pennsylvania'] # pick out pennsylvania information
  l_budget_deficit = max(q4_list, key = lambda x:x[1])  #find the largest one 
  return print(l_budget_deficit[0],'in Pennsylvania had the largest budget deficit and the amount is',l_budget_deficit[1]*1000,'dollars')

In [ ]:
q4_biggest_pennsylvania_district(districts_list)

PENN HILLS SD in Pennsylvania had the largest budget deficit and the amount is 32789000 dollars


**Conclusion:** ***PENN HILLS SD*** in Pennsylvania had the largest budget deficit. And the amout is ***32,789,000 dollars***.

## Q5

New England comprises the states Maine, Vermont, New Hampshire, Rhode Island, Massachusetts, and Connecticut. To calculate per capita statistics, you divide the total amount by the total population size. Which state in New England gave the most non-federal money (state and local combined) to schools? How much did they give per capita?

In [ ]:
def q5_new_england_highest_per_capita(districts_list, state_populations):
# calculate the total non-federal money of six states
  states_list = ['Maine','Vermont','New Hampshire','Rhode Island','Massachusetts','Connecticut']
  non_federal = {}  # create a list to store non-federal money of six states
  for item in districts_list:
    if item['state'] in states_list:
      non_federal[item['state']] = non_federal.get(item['state'],0) + item['local_revenue'] + item['state_revenue']     
    else:
      pass

  # calculate the per capita of every state
  per_capita = {}
  for key in state_populations:
    if non_federal.get(key):
      per_capita[key] = non_federal[key] / state_populations[key]

  # find out the smallest mean size
  for key,value in non_federal.items():
      if(value == max(non_federal.values())):
          result = [key,value]
  return print(result[0],'in New England gave the most non-federal money to school. They gave',round((per_capita.get(result[0])*1000),2),'dollars per capita.')

In [ ]:
q5_new_england_highest_per_capita(districts_list, state_populations)

Massachusetts in New England gave the most non-federal money to school. They gave 2498.29 dollars per capita.


**Conclusion:** ***Massachusetts*** in New England gave the most non-federal money to school. They gave ***2498.29 dollars*** per capita.

## Q6

The mean size of a school district in a state is equal to the total population divided by the total number of school districts. Which state had the smallest mean school district size? How many people, on average, live in each school district in that state? 

In [ ]:
def q6_smallest_mean_district_size(districts_list, state_populations):
# count the number of school stricts in every state
  state =[]
  for item in districts_list:
    state.append(item['state'])

  state_cnt = {}
  for i in state:
    state_cnt[i] = state_cnt.get(i,0) + 1 
  
# calculate the mean size of every state
  mean_size = {}
  for key in state_populations:
    if state_cnt.get(key):
      mean_size[key] = state_populations[key] / state_cnt[key]

# find out the smallest mean size
  for key,value in mean_size.items():
      if (value == min(mean_size.values())):
          result = [key,value]
  return print(result[0],'had the smallest mean school district size. About',int(result[1]),'people on average live in each school district in the state.')

In [ ]:
q6_smallest_mean_district_size(districts_list, state_populations)

Vermont had the smallest mean school district size. About 1955 people on average live in each school district in the state.


**Conclusion:** ***Vermont*** had the smallest mean school district size. About ***1955 people*** on average live in each school district in the state.

# Part 3



## Q1 

Say we are interested in judging which state receives the “best deal” in federal aid for schools. Decide on two different metrics to measure how much relative support a state receives (for instance, relative to its population or in comparison to other states). 


***The first metric is the proportion of each state's federal funding in total funding.*** It is equal to the federal_revenue divided by revenue in the dataset of district_list. The metric can be used to descibe the funding components from all levels of governments and to compare which state had the largest prportion of funding from federal aid. The larger the number is, the relatively more support the state received from federal governments compared to others.

***The second metric is the ratio of state's per capita to the federal average level.*** It is equal to each state's per capita divided by federal's per capita. If the ratio is greater than 1, it means that the state's per capita is higher than the federal average. The larger the ratio is, the more federal aid people received. We are going to find the largest ratio.

## Q2

Implement each metric you described above and generate a sorted (ranked) list of states from "best" to "worst" financial deal from the federal government. Your submission should include the ranked list for each metric as output from your code.


In [ ]:
# Implement your first metric here.
# Metric 1 is the proportion of each state's federal funding in total funding 
# And the formula is sum(federal_revenue) / sum(revenue)
def metric_1(districts_list):
  total_revenue = {} # calculate total revenue of each state
  total_federal_revenue = {} # calculate total federal_revenue of each state
  for item in districts_list:
    total_revenue[item['state']] = total_revenue.get(item['state'],0) + item['revenue']
    total_federal_revenue[item['state']] = total_federal_revenue.get(item['state'],0) + item['federal_revenue']

  # calculate the proportion
  federal_revenue_proportion = {}
  for key in total_revenue:
    if total_federal_revenue.get(key):
      federal_revenue_proportion[key] = (total_federal_revenue[key] / total_revenue[key]) * 100  # with the value in percentage(%) format

  return sorted(federal_revenue_proportion.items(), key = lambda item:item[1],reverse = True) # rank the results from high to low

In [ ]:
metric_1(districts_list)

[('Louisiana', 15.181995445749402),
 ('Mississippi', 14.84050797041844),
 ('South Dakota', 13.838318226545338),
 ('New Mexico', 12.718679727102947),
 ('Arizona', 12.38279980581979),
 ('Alaska', 12.11460943031492),
 ('Florida', 11.933780212200098),
 ('Tennessee', 11.8575757687289),
 ('Montana', 11.646066245399625),
 ('Oklahoma', 11.429327845829887),
 ('Kentucky', 11.421277073668263),
 ('North Carolina', 11.149171945246405),
 ('Idaho', 11.025019408780121),
 ('Alabama', 10.747158504839478),
 ('Texas', 10.712524685907143),
 ('Arkansas', 10.656251659547996),
 ('Hawaii', 10.642330434073198),
 ('California', 10.241354663662186),
 ('Georgia', 10.070846949713603),
 ('West Virginia', 10.013317632505608),
 ('District of Columbia', 9.92591982208876),
 ('North Dakota', 9.917961575686448),
 ('South Carolina', 9.581302428100136),
 ('Nevada', 8.919143544537988),
 ('Missouri', 8.795716805561321),
 ('Michigan', 8.766084905672264),
 ('Utah', 8.68231377448409),
 ('Rhode Island', 8.085081044232451),
 ('Neb

In [ ]:
# Implement your second metric here.
# Metric 2 is the ratio of state's per capita to federal's average per capita
# And the formula is per capita of each state / federal's per capita
# federal's per capita = sum(federal_revenue) / sum(population)
def metric_2(districts_list,state_populations):
  # calculate the federal's per capita
    # calculate the total federal_revenue
  total_federal = {}
  for item in districts_list:
    total_federal[item['state']] = total_federal.get(item['state'],0) + item['federal_revenue'] 
  sum_federal_revenue = 0
  for i in total_federal:
    sum_federal_revenue = sum_federal_revenue + total_federal[i]

   # calculate the totol populations
  sum_populations = 0
  for i in state_populations: # ignore the population statistis of Puerto Rico, because this state has no funding data
    if i != 'Puerto Rico':
      sum_populations = sum_populations + state_populations[i]

    # calculate the average per capita of the whole federal
  per_capita_federal = sum_federal_revenue / sum_populations * 1000
  
  # calculate the per capita of each state
  federal_revenue_each_state = {}  # create a list to store federal revenue each state
  for item in districts_list:
    federal_revenue_each_state[item['state']] = federal_revenue_each_state.get(item['state'],0) + item['federal_revenue']     

    # calculate the per capita of every state
  per_capita_each_state = {}
  for key in state_populations:
    if federal_revenue_each_state.get(key):
      per_capita_each_state[key] = federal_revenue_each_state[key] / state_populations[key] * 1000

    # calculate the ratio
  ratio = {k:v/per_capita_federal for (k,v) in per_capita_each_state.items()}

  return sorted(ratio.items(), key = lambda item:item[1], reverse = True) # rank the results from high to low

In [ ]:
metric_2(districts_list,state_populations)

[('Alaska', 2.4845520806924695), ('District of Columbia', 1.9629890941800043), ('Louisiana', 1.7060599835297314), ('Texas', 1.3296133316145875), ('North Dakota', 1.3237358670939119), ('New Mexico', 1.31190518151694), ('South Dakota', 1.3037937991186908), ('Mississippi', 1.2780536817963257), ('Hawaii', 1.2027648211940662), ('Montana', 1.1596492708540387), ('Wyoming', 1.1400386483892913), ('California', 1.134839984769168), ('Arkansas', 1.0984983126941943), ('Kentucky', 1.0848507404768384), ('West Virginia', 1.082848230078282), ('Rhode Island', 1.0663545877609069), ('Georgia', 1.0640749061019255), ('Oklahoma', 1.0510370466410792), ('Arizona', 1.0460607766789816), ('Illinois', 1.0228139125906557), ('South Carolina', 1.0198093116330047), ('Michigan', 1.003248633108739), ('Nebraska', 0.9932019093763534), ('New York', 0.9866241945621059), ('Tennessee', 0.9840300501561009), ('Alabama', 0.9513165774941506), ('Vermont', 0.9464682472520418), ('Florida', 0.9436419577005775), ('North Carolina', 0.9

## Q3


***I recommend the second metric to rank states*** on how much federal financial aid they receive relative to other states. The main reason is that the metric 2 considers the population.

The second metric represents the financial federal aid every person can receive in the state compared to the federal average level and it takes populations of states into consideration. Take the state of Pennsylvania as an example, the per capita is 151.76 dollars. The average per capita of the whole federal is 175.41 dollars. So the ratio is around 0.87. This number is smaller than 1, which represents that compared to the national level, people in Pennsylvania receive less. The second metric uses such method to make comparsions among states.

Compared to Metric 1, I think Metric 2 has the following advantages:


*   ***It contains both revenues and population data.*** Metric 1 only considers revenues regardless of the number of people living in the area. Since the beneficiary of the financial aid is citizens, we should take population sizes into consideration when making comparisons.
*   ***It weakens the effect of gaps in economic development levels in each state and the impact of the local people's enthusiasm for joining in education funding.*** When using Metric 1, the proportion of federal funding is linked with the other two parts - local funding and state funding. In some economically developed states, the state government may invest more money in education. And some states may have many citizens who are more willing to put their money in education. The increasing in these numbers will reduce the proportion of federal funding, but it does not mean that the federal supports less on the state. Metric 2 can avoid this situation.

So, I think Metric 2 is more appropriate to rank states in this set of questions.



# Extra Credit


I want to explore more in two questions. ***The first one is the share of each state's education budget in the state's total annual financial budget, which gives me ideas on the importance of developing education from the perspective of government finance. The second one is the components of expenses, then finding out which parts of education spend more.***

The first questiuon is to study the proportion of education budget in the total budget. Through the proportions, we can compare the distribution of government budgets in different states. The question is interesting because we can understand the relative importance of education through the perspective of budget allocation. Data analysts can compare budget allocations in different fields, and then assess wherther the budget is adequate according to the actual expenditures, and whether it is necessary to further increase the budget in the future development. 

The second question is to study the expenditure details of education expenses in public school districts and states. I think this question is interesting because it helps me have a basic understanding of the expenses structure and understand which parts pf education cost more than others. Data analysts can further analyze the cost details, understand the efficiency of capital use and budget priorities, and then make targeted policy recommendations, such as where to increase inputs and how to improve capital efficiency.

For the first question, we have had funding and expenses data. We still need the total budget of each state and the budget in other public areas, such as infrastructure budget, culture affairs budget, technology budget and so on. We can easily use proportion of education budget in each state to proceed the analysis. And the proportion of education budget is equal to state's education budget divided by total annual budget of the state. Through the ratio, we can rank the states to find out states which spend more on education than others. Also, we judge whether the current budget can meet the needs and how to make the following policies.

For the second question, now we only have the total expenses of each public school district. We need more detailed data to conduct analyses. The existd data cannot meet our needs. Key components of expenses, such as faculty expenditure, facilities expenditure, expenses on different levels of education period, such as primary, secondary, high school, etc. From the data, we can easily get the proportion by using the division formula. When we get the results, we can compare school districts and states, and observe the differences and key parts of expenditure in each state. 